src/APIIndex.php

```diff
@@ -0,0 +1,71 @@
+<?hh
+
+namespace HHVM\UserDocumentation;
+
+class APIIndex {
+  private static function getIndex(
+  ): Map<string, Map<string, APIClassIndexEntry>> {
+    return require(BuildPaths::APIDOCS_INDEX);
+  }
+
+  public static function getProducts(): ImmVector<string> {
+    return self::getIndex()->keys()->toImmVector();
+  }
+  
+  public static function getReferenceForType(
+    string $type,
+  ): Map<string, APIClassIndexEntry> {
+    $index = self::getIndex();
+    invariant(
+      $index->containsKey($type),
+      '%s is not in the index',
+      $type,
+    );
+    return $index[$type];
+  }
+
+  public static function getFileForAPI(
+    string $type,
+    string $api,
+    ?string $method,
+  ) {
+    $index = self::getIndex();
+    invariant(
+      $index->containsKey($type),
+      'Type %s does not exist',
+      $type,
+    );
+    invariant(
+      $index[$type]->containsKey($api),
+      'Type %s does not contain reference %s',
+      $type,
+      $api,
+    );
+    invariant(
+      isset($index[$type][$api]['path']),

```

Comment:
I tried using ->containsKey and got typing errors, but array_key_exists might work